src/main/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluator.java

```diff
@@ -0,0 +1,130 @@
+import org.opensearch.security.configuration.ClusterInfoHolder;
+import org.opensearch.security.securityconf.ConfigModel;
+import org.opensearch.security.securityconf.DynamicConfigModel;
+import org.opensearch.security.securityconf.SecurityRoles;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+
+public class RestLayerPrivilegesEvaluator {
+    protected final Logger log = LogManager.getLogger(this.getClass());
+    private final ClusterService clusterService;
+    private final AuditLog auditLog;
+    private ThreadContext threadContext;
+    private final ClusterInfoHolder clusterInfoHolder;
+    private ConfigModel configModel;
+    private DynamicConfigModel dcm;
+    private final AtomicReference<NamedXContentRegistry> namedXContentRegistry;
+
+    public RestLayerPrivilegesEvaluator(
+        final ClusterService clusterService,
+        final ThreadPool threadPool,
+        AuditLog auditLog,
+        final ClusterInfoHolder clusterInfoHolder,
+        AtomicReference<NamedXContentRegistry> namedXContentRegistry
+    ) {
+        this.clusterService = clusterService;
+        this.auditLog = auditLog;
+
+        this.threadContext = threadPool.getThreadContext();
+
+        this.clusterInfoHolder = clusterInfoHolder;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    @Subscribe
+    public void onConfigModelChanged(ConfigModel configModel) {
+        this.configModel = configModel;
+    }
+
+    @Subscribe
+    public void onDynamicConfigModelChanged(DynamicConfigModel dcm) {
+        this.dcm = dcm;
+    }
+
+    private SecurityRoles getSecurityRoles(Set<String> roles) {
+        return configModel.getSecurityRoles().filter(roles);
+    }
+
+    public boolean isInitialized() {
+        return configModel != null && configModel.getSecurityRoles() != null && dcm != null;
+    }
+
+    public PrivilegesEvaluatorResponse evaluate(final User user, Set<String> action0) {
+
+        if (!isInitialized()) {
+            throw new OpenSearchSecurityException("OpenSearch Security is not initialized.");
+        }
+
+        final PrivilegesEvaluatorResponse presponse = new PrivilegesEvaluatorResponse();
+
+        final TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
+
+        Set<String> mappedRoles = mapRoles(user, caller);
+
+        presponse.resolvedSecurityRoles.addAll(mappedRoles);
+        final SecurityRoles securityRoles = getSecurityRoles(mappedRoles);
+
+        final boolean isDebugEnabled = log.isDebugEnabled();
+        if (isDebugEnabled) {
+            log.debug("Evaluate permissions for {} on {}", user, clusterService.localNode().getName());
+            log.debug("Action: {}", action0);
+            log.debug("Mapped roles: {}", mappedRoles.toString());
+        }
+
+        for (String action : action0) {
```

Comment:
nit: I had to reference up above that `action0` is a set. This looks like its looping through a singular action.

Reply:
It will be singular for `name`, but it will be in loop for `actionNames`